src/backend/distributed/operations/shard_split.c

```diff
@@ -0,0 +1,1117 @@
+		 * Foreign key constraints are created after Metadata changes (see CreateForeignKeyConstraints).
+		 */
+		CreateAndCopySplitShardsForShardGroup(
+			mapOfShardToPlacementCreatedByWorkflow,
+			sourceShardToCopyNode,
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			workersForPlacementList);
+
+		/*
+		 * Drop old shards and delete related metadata. Have to do that before
+		 * creating the new shard metadata, because there's cross-checks
+		 * preventing inconsistent metadata (like overlapping shards).
+		 */
+		DropShardList(sourceColocatedShardIntervalList);
+
+		/* Insert new shard and placement metdata */
+		InsertSplitChildrenShardMetadata(shardGroupSplitIntervalListList,
+										 workersForPlacementList);
+
+		/*
+		 * Create foreign keys if exists after the metadata changes happening in
+		 * DropShardList() and InsertSplitChildrenShardMetadata() because the foreign
+		 * key creation depends on the new metadata.
+		 */
+		CreateForeignKeyConstraints(shardGroupSplitIntervalListList,
+									workersForPlacementList);
+	}
+	PG_CATCH();
+	{
+		/* Do a best effort cleanup of shards created on workers in the above block */
+		TryDropSplitShardsOnFailure(mapOfShardToPlacementCreatedByWorkflow);
+
+		PG_RE_THROW();
+	}
+	PG_END_TRY();
+
+	CitusInvalidateRelcacheByRelid(DistShardRelationId());
+}
+
+
+/* Create ShardGroup split children on a list of corresponding workers. */
+static void
+CreateSplitShardsForShardGroup(HTAB *mapOfShardToPlacementCreatedByWorkflow,
+							   List *shardGroupSplitIntervalListList,
+							   List *workersForPlacementList)
+{
+	/*
+	 * Iterate over all the shards in the shard group.
+	 */
+	List *shardIntervalList = NIL;
+	foreach_ptr(shardIntervalList, shardGroupSplitIntervalListList)
+	{
+		ShardInterval *shardInterval = NULL;
+		WorkerNode *workerPlacementNode = NULL;
+
+		/*
+		 * Iterate on split shards DDL command list for a given shard
+		 * and create them on corresponding workerPlacementNode.
+		 */
+		forboth_ptr(shardInterval, shardIntervalList, workerPlacementNode,
+					workersForPlacementList)
+		{
+			/* Populate list of commands necessary to create shard interval on destination */
+			List *splitShardCreationCommandList = GetPreLoadTableCreationCommands(
+				shardInterval->relationId,
+				false, /* includeSequenceDefaults */
+				NULL /* auto add columnar options for cstore tables */);
+			splitShardCreationCommandList = WorkerApplyShardDDLCommandList(
+				splitShardCreationCommandList,
+				shardInterval->shardId);
+
+			/* Create new split child shard on the specified placement list */
+			CreateObjectOnPlacement(splitShardCreationCommandList, workerPlacementNode);
+
+			ShardCreatedByWorkflowEntry entry;
+			entry.shardIntervalKey = shardInterval;
+			entry.workerNodeValue = workerPlacementNode;
+			bool found = false;
+			hash_search(mapOfShardToPlacementCreatedByWorkflow, &entry, HASH_ENTER,
+						&found);
```

Comment:
it might be slightly better to enter before creation, in case e.g. this command throws an out of memory error.

Reply:
If we insert into HashTable before, the problem is that if a table with same name exists and we fail because of that, we will end up deleting it as well.